File: src/main/java/com/teamwizardry/librarianlib/sprite/SpritesMetadataSection.java
package com.teamwizardry.librarianlib.sprite;

import java.util.List;

import net.minecraft.client.resources.data.IMetadataSection;

public class SpritesMetadataSection implements IMetadataSection {

	public List<SpriteDefinition> definitions;
	public int width, height;
	
	public SpritesMetadataSection(int width, int height, List<SpriteDefinition> definitions) {
		super();
		this.width = width;
		this.height = height;
		this.definitions = definitions;
	}

	public static class SpriteDefinition {
		public String name;
		public int u, v, w, h, offsetU, offsetV;
		public int[] frames;
		
		public SpriteDefinition(String name, int u, int v, int w, int h, int[] animationTimes, int offsetU, int offsetV) {
			super();
			this.name = name;
			this.u = u;
			this.v = v;
			this.w = w;
			this.h = h;
			this.frames = animationTimes;
			this.offsetU = offsetU;
			this.offsetV = offsetV;
		}
	}
	
}

File: modules/sprite/src/main/java/com/teamwizardry/librarianlib/sprite/SpritesMetadataSection.kt
package com.teamwizardry.librarianlib.sprite

import net.minecraft.client.resources.data.IMetadataSection
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import java.util.Arrays

@SideOnly(Side.CLIENT)
class SpritesMetadataSection(var width: Int, var height: Int, var sprites: List<SpriteDefinition>, var colors: List<ColorDefinition>) : IMetadataSection {
    companion object {
        var registered = false
    }
}

class SpriteDefinition(var name: String) {

    var u: Int = 0
    var v: Int = 0
    var w: Int = 0
    var h: Int = 0

    var frames: IntArray = IntArray(0)
    var offsetU: Int = 0
    var offsetV: Int = 0

    var minUCap: Int = 0
    var minVCap: Int = 0
    var maxUCap: Int = 0
    var maxVCap: Int = 0

    var pinTop: Boolean = true
    var pinRight: Boolean = true
    var pinBottom: Boolean = true
    var pinLeft: Boolean = true

    override fun toString(): String {
        return "SpriteDefinition(" +
            "name='$name', uv=($u,$v), wh=($w,$h), frames=${Arrays.toString(frames)}, offsetUV=($offsetU,$offsetV), " +
            "cap=($minUCap,$minVCap,$maxUCap,$maxVCap), pinned=($pinTop,$pinRight,$pinBottom,$pinLeft)" +
            ")"
    }
}

class ColorDefinition(var name: String, var u: Int, var v: Int) {
    constructor(name: String) : this(name, -1, -1)
}
